src/java/org/apache/cassandra/db/virtual/PrimaryIdTable.java

```diff
@@ -0,0 +1,350 @@
+                    long size = current - lastPosition;
+
+                    String keyString = target.partitionKeyType.getString(key.getKey());
+
+                    // Check if the current key is outside the queried range; if so, stop
+                    if (range.right.compareTo(key) < 0)
+                        return endOfData();
+
+                    // Convert the token to a string and create a clustering object
+                    String tokenString = key.getToken().toString();
+                    Clustering<?> clustering = Clustering.make(
+                        IntegerType.instance.decompose(new BigInteger(tokenString)),
+                        UTF8Type.instance.decompose(keyString)
+                    );
+
+                    // Check if the current clustering matches the filter; if so, return the row
+                    if (filter.selects(clustering))
+                        return buildRow(clustering, size);
+                }
+                return endOfData();
+            }
+        };
+    }
+
+    /**
+     * This converts the clustering token/key into the partition level token/key for the target table. Also provides an
+     * optimization from RowFilter when a `key` is specified with or without the clustering `token` being set.
+     */
+    private AbstractBounds<PartitionPosition> getBounds(TableMetadata target, ClusteringIndexFilter clusteringIndexFilter, RowFilter rowFilter)
+    {
+        Slices s = clusteringIndexFilter.getSlices(metadata);
+        Token startToken = target.partitioner.getMinimumToken();
+        Token endToken = target.partitioner.getMaximumToken();
+
+        // find min/max token values from the clustering key
+        for (int i = 0; i < s.size(); i++)
+        {
+            Slice slice = s.get(i);
+            if (!slice.start().isEmpty())
+            {
+                BigInteger token = IntegerType.instance.compose(slice.start().bufferAt(0));
+                startToken = target.partitioner.getTokenFactory().fromString(token.toString());
+            }
+            if (!slice.end().isEmpty())
+            {
+                BigInteger token = IntegerType.instance.compose(slice.end().bufferAt(0));
+                endToken = target.partitioner.getTokenFactory().fromString(token.toString());
+            }
+        }
+
+        // override min/max of token if the `key` is specified
+        for (RowFilter.Expression expression : rowFilter.getExpressions())
+        {
+            if (expression.column().name.toString().equals(COLUMN_KEY))
+            {
+                if (expression.operator() != Operator.EQ)
+                    throw new InvalidRequestException(KEY_ONLY_EQUALS_ERROR);
+
+                String keyString = UTF8Type.instance.compose(expression.getIndexValue());
+                ByteBuffer keyAsBB = target.partitionKeyType.fromString(keyString);
```

Comment:
```suggestion
                String keyString = UTF8Type.instance.compose(expression.getIndexValue());
                ByteBuffer keyAsBB = target.partitionKeyType.asCQL3Type().fromCQLLiteral(keyString);
```

can we use cql literals rather than type strings?  type strings are not the same and are not the best for human readability.

Reply:
commented above, its for consistency between output/input since we dont have anything to convert CQL literals -> type (that i know of at least)